climada_petals/engine/supplychain.py

```diff
@@ -19,226 +19,479 @@
+       production from a Multi-Regional Input-Output Table dataframe.
+
+    Parameters
+    ----------
+    v : pandas.DataFrame or numpy.array
+        a row vector of the total final added-value
+    G : pandas.DataFrame or numpy.array
+        Symmetric input output Ghosh table
+    """
+
+    start_row, end_row = rows_data
+    start_col, end_col = cols_data
+
+    sectors = mriot_df.iloc[start_row:end_row, col_sectors].unique()
+    regions = mriot_df.iloc[start_row:end_row, col_iso3].unique()
+    multiindex = pd.MultiIndex.from_product(
+        [regions, sectors], names=["region", "sector"]
+    )
+
+    Z = mriot_df.iloc[start_row:end_row, start_col:end_col].values.astype(float)
+    Z = pd.DataFrame(data=Z, index=multiindex, columns=multiindex)
+
+    Y = mriot_df.iloc[start_row:end_row, end_col:-1].sum(1).values.astype(float)
+    Y = pd.DataFrame(data=Y, index=multiindex, columns=["final demand"])
+
+    x = mriot_df.iloc[start_row:end_row, -1].values.astype(float)
+    x = pd.DataFrame(data=x, index=multiindex, columns=["total production"])
+
+    return Z, Y, x
+
+
+def calc_v(Z, x):
+    """Calculate value added (v) from Z and x
+
+    value added = industry output (x) - inter-industry inputs (sum_rows(Z))
+
+    Parameters
+    ----------
+    Z : pandas.DataFrame or numpy.array
+        Symmetric input output table (flows)
+    x : pandas.DataFrame or numpy.array
+        industry output
+
+    Returns
+    -------
+    pandas.DataFrame or numpy.array
+        Value added v as row vector
+    """
+
+    value_added = np.diff(np.vstack((Z.sum(0), x.T)), axis=0)
+    if isinstance(Z, pd.DataFrame):
+        value_added = pd.DataFrame(value_added, columns=Z.index, index=["indout"])
+    if isinstance(value_added, pd.Series):
+        value_added = pd.DataFrame(value_added)
+    if isinstance(value_added, pd.DataFrame):
+        value_added.index = ["indout"]
+    return value_added
+
+
+def calc_B(Z, x):
+    """Calculate the B matrix (allocation coefficients matrix)
+    from Z matrix and x vector
+
+    Parameters
+    ----------
+    Z : pandas.DataFrame or numpy.array
+        Symmetric input output table (flows)
+    x : pandas.DataFrame or numpy.array
+        Industry output column vector
+
+    Returns
+    -------
+    pandas.DataFrame or numpy.array
+        Symmetric input output table (allocation matrix) B
+        The type is determined by the type of Z.
+        If DataFrame index/columns as Z
+
+    Notes
+    -----
+    This function adapts pymrio.tools.iomath.calc_A to compute
+    the allocation coefficients matrix B.
+    """
+
+    if isinstance(x, (pd.DataFrame, pd.Series)):
+        x = x.values
+    if (type(x) is not np.ndarray) and (x == 0):
+        recix = 0
+    else:
+        with warnings.catch_warnings():
+            # catch the divide by zero warning
+            # we deal wit that by setting to 0 afterwards
+            warnings.simplefilter("ignore")
```

Comment:
Only ignore the divide by zero warning
```suggestion
            # Ignore devide by zero warning, we set to 0 afterwards
            warnings.filterwarnings("ignore", message="divide by zero")
```